shared/common-adapters/list2.native.js

```diff
@@ -0,0 +1,81 @@
+// @flow
+import React, {PureComponent} from 'react'
+import {FlatList, View} from 'react-native'
+import {globalStyles, collapseStyles, styleSheetCreate} from '../styles'
+import {memoize} from '../util/memoize'
+
+import type {Props} from './list2'
+
+class List2 extends PureComponent<Props<any>, void> {
+  static defaultProps = {
+    keyboardShouldPersistTaps: 'handled',
+  }
+  _itemRender = ({item, index}) => {
+    return this.props.renderItem(index, item)
+  }
+
+  _memoizedOffsetGetter = memoize((data, index) => {
+    if (index === 0) {
+      return 0
+    }
+    if (!index) {
+      return 0
+    }
+    // $FlowIssue can't figure out type for data.
+    return this._memoizedOffsetGetter(index - 1) + this.props.itemHeight(index, data[index])
+  })
+
+  _getItemLayout = (data, index) => {
+    if (typeof this.props.itemHeight === 'number') {
+      return {
+        index,
+        length: this.props.itemHeight,
+        offset: this.props.itemHeight * index,
+      }
+    }
+    return {
+      index,
+      length: this.props.itemHeight(index, data[index]),
+      offset: this._memoizedOffsetGetter(data, index),
```

Comment:
why is this memozied like this? every call it'll get thrown away

Reply:
additionally i dont think this will be required to be called in order so i don't think you can generalize this

Reply:
I did some simple test and it was going through the array from beginning to end. So memoization like this means we're always calculating the offset based on the previous offset, which saves us from going from the beginning the array everytime. Why is it thrown away?

Reply:
that won't always be true, it can skip things depending on how you scroll etc

Reply:
note this is a memoize-one, aka the last value is saved so every item it'll get thrown away

Reply:
Even if it skips, it still works. For example, if it memoized index=2, then suddently skipped to 4. We'd try to find 4, which isn't memoized, which would then look for 3, which isn't there either, causing it to look for 2, which exists. All these are recursive, so non of them would override the memoized value until they return. So we'd always benefit from the latest memoized value.

It won't help if it's going backwards though, which I guess can only be solved by having a dictionary memoizing all indices. Would you prefer that?